File: 235-lowest-common-ancestor-of-a-binary-search-tree/lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        TreeNode* cur = root;
        while(cur) {
            cout << cur->val << ' ';
            if(p->val > cur->val && q->val > cur->val) {
                // search right
                cur = cur->right;
            }
            else if(p->val < cur->val && q->val < cur->val) {
                // search left
                cur = cur->left;
            }
            else {
                return cur;
            }
        }
        return cur;
    }
};

File: 235-lowest-common-ancestor-of-a-binary-search-tree/lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        par = {root: None}
        qu = deque([root])
        visited = set([root])
        while q not in visited or p not in visited:
            cur = qu.popleft()
            if cur.left:
                qu.append(cur.left)
                visited.add(cur.left)
                par[cur.left] = cur
            if cur.right:
                qu.append(cur.right)
                visited.add(cur.right)
                par[cur.right] = cur
        
        p_path = []
        cur = p
        while cur:
            p_path.append(cur)
            cur = par[cur]
        
        cur = q
        while cur not in p_path:
            cur = par[cur]
        
        return cur
